presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PullUpExpressionInLambdaRules.java

```diff
@@ -0,0 +1,348 @@
+                boolean allArgumentsValid = validRowExpressionMap.values().stream().allMatch(x -> x.equals(Boolean.TRUE));
+                if (!allArgumentsValid) {
+                    candidates.addAll(validRowExpressionMap.entrySet().stream()
+                            .filter(x -> x.getValue().equals(Boolean.TRUE))
+                            .filter(x -> x.getKey() instanceof CallExpression || x.getKey() instanceof SpecialFormExpression)
+                            .map(Map.Entry::getKey)
+                            .collect(toImmutableList()));
+                }
+                return allArgumentsValid && determinismEvaluator.isDeterministic(call);
+            }
+            return false;
+        }
+
+        @Override
+        public Boolean visitSpecialForm(SpecialFormExpression specialForm, Boolean context)
+        {
+            // Bind expression will complicate the lambda expression, we apply this optimization before DesugarLambdaRule. And if there are bind expression, skip
+            if (specialForm.getForm().equals(SpecialFormExpression.Form.BIND)) {
+                return false;
+            }
+            Map<RowExpression, Boolean> validRowExpressionMap = specialForm.getArguments().stream().distinct().collect(toImmutableMap(identity(), x -> x.accept(this, context)));
+            if (context.equals(Boolean.TRUE)) {
+                boolean allArgumentsValid = validRowExpressionMap.values().stream().allMatch(x -> x.equals(Boolean.TRUE));
+                if (!allArgumentsValid) {
+                    candidates.addAll(validRowExpressionMap.entrySet().stream()
+                            .filter(x -> x.getValue().equals(Boolean.TRUE))
+                            .filter(x -> x.getKey() instanceof CallExpression || x.getKey() instanceof SpecialFormExpression)
+                            .map(Map.Entry::getKey)
+                            .collect(toImmutableList()));
+                }
+                return allArgumentsValid && determinismEvaluator.isDeterministic(specialForm);
+            }
+            return false;
+        }
+
+        @Override
+        public Boolean visitLambda(LambdaDefinitionExpression lambda, Boolean context)
+        {
+            if (lambda.getBody().accept(this, true)) {
+                candidates.add(lambda.getBody());
+            }
+            return false;
+        }
+
+        @Override
+        public Boolean visitVariableReference(VariableReferenceExpression reference, Boolean context)
+        {
+            return inputVariables.contains(reference);
+        }
+
+        @Override
+        public Boolean visitConstant(ConstantExpression literal, Boolean context)
+        {
+            return true;
+        }
+
+        @Override
+        public Boolean visitInputReference(InputReferenceExpression reference, Boolean context)
+        {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, "Unexpected input reference in expression");
```

Comment:
This is opportunistic optimization. Lets not crash. Maybe make it a warning return from the optimizer?

Reply:
We should not have InputReferenceExpression at this stage. Removed the exception though.